mlir/lib/Quantum/Transforms/ChainedHadamard.cpp

```diff
@@ -0,0 +1,86 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "chained-hadamard"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Quantum/IR/QuantumInterfaces.h"
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+#include "Quantum/Utils/QuantumSplitting.h"
+
+using llvm::dbgs;
+using namespace mlir;
+using namespace catalyst;
+using namespace catalyst::quantum;
+
+namespace {
+
+struct ChainedHadamardOpRewritePattern : public mlir::OpRewritePattern<CustomOp> {
```

Comment:
The pattern looks great 💯

Reply:
Out of curiosity, are you familiar with other ways this pattern/rewrite could be specified, and if so how would that look?

Reply:
Thanks, these are good observations, although like you mention I don't think they would necessarily offer any improvements. I was thinking more along the lines of specifying the transform with DRR or PDLL, or other mechanism offered by MLIR :)

Reply:
Oh right! DRR would work but I'm not a fan of declarative rewrites just because the language isn't as readable or intuitive. PDLL has similar issues for me personally. I'm not entirely well versed with either of those but have read PDLL for debugging purposes. Something I previously worked with was a pattern rewrite library in python that used Python bindings for PDLL which solved all these issues for me!

Reply:
Interesting, I'd be curious to hear about the second approach. For now I'll resolve this thread :)